Switching channels

fxedel opened the pull request:
See #108.

FloEdelmann commented:
I like our tests :)

Very nice that you start implementing this right away!

FloEdelmann commented:
<!-- github-plugin-diff -->
<!-- commit = a218290800e6f921325d7ebea6a499570587afbb -->
Last updated: 5/3/2017, 9:38:26 PM
## Diff plugin outputs test
### Modified plugin `qlcplus` in this PR
Plugins are always tested with the following fixtures: fixtures/cameo/thunder-wash-600-w.json, fixtures/lightmaxx/vega-zoom-wash.json
```diff
Index: cameo/thunder-wash-600-w.qxf
===================================================================
--- cameo/thunder-wash-600-w.qxf
+++ cameo/thunder-wash-600-w.qxf
@@ -77,46 +77,46 @@
   <Capability Min="0" Max="255">0-100%</Capability>
  </Channel>
  <Mode Name="1-channel">
   <Physical>
-   <Bulb ColourTemperature="0" Type="LED" Lumens="0" />
-   <Dimensions Width="335" Height="182" Depth="93" Weight="3.25" />
-   <Lens Name="Other" DegreesMin="75" DegreesMax="75" />
-   <Focus Type="Fixed" TiltMax="0" PanMax="0" />
-   <Technical DmxConnector="3-pin" PowerConsumption="130" />
+   <Bulb ColourTemperature="0" Type="LED" Lumens="0"/>
+   <Dimensions Width="335" Height="182" Depth="93" Weight="3.25"/>
+   <Lens Name="Other" DegreesMin="75" DegreesMax="75"/>
+   <Focus Type="Fixed" TiltMax="0" PanMax="0"/>
+   <Technical DmxConnector="3-pin" PowerConsumption="130"/>
   </Physical>
   <Channel Number="0">Strobe</Channel>
  </Mode>
  <Mode Name="2-channel">
   <Physical>
-   <Bulb ColourTemperature="0" Type="LED" Lumens="0" />
-   <Dimensions Width="335" Height="182" Depth="93" Weight="3.25" />
-   <Lens Name="Other" DegreesMin="75" DegreesMax="75" />
-   <Focus Type="Fixed" TiltMax="0" PanMax="0" />
-   <Technical DmxConnector="3-pin" PowerConsumption="130" />
+   <Bulb ColourTemperature="0" Type="LED" Lumens="0"/>
+   <Dimensions Width="335" Height="182" Depth="93" Weight="3.25"/>
+   <Lens Name="Other" DegreesMin="75" DegreesMax="75"/>
+   <Focus Type="Fixed" TiltMax="0" PanMax="0"/>
+   <Technical DmxConnector="3-pin" PowerConsumption="130"/>
   </Physical>
   <Channel Number="0">Dimmer</Channel>
   <Channel Number="1">Strobe 2</Channel>
  </Mode>
  <Mode Name="3-channel">
   <Physical>
-   <Bulb ColourTemperature="0" Type="LED" Lumens="0" />
-   <Dimensions Width="335" Height="182" Depth="93" Weight="3.25" />
-   <Lens Name="Other" DegreesMin="75" DegreesMax="75" />
-   <Focus Type="Fixed" TiltMax="0" PanMax="0" />
-   <Technical DmxConnector="3-pin" PowerConsumption="130" />
+   <Bulb ColourTemperature="0" Type="LED" Lumens="0"/>
+   <Dimensions Width="335" Height="182" Depth="93" Weight="3.25"/>
+   <Lens Name="Other" DegreesMin="75" DegreesMax="75"/>
+   <Focus Type="Fixed" TiltMax="0" PanMax="0"/>
+   <Technical DmxConnector="3-pin" PowerConsumption="130"/>
   </Physical>
   <Channel Number="0">Dimmer</Channel>
   <Channel Number="1">Strobe 2</Channel>
   <Channel Number="2">Flash Duration</Channel>
  </Mode>
  <Mode Name="5-channel">
   <Physical>
-   <Bulb ColourTemperature="0" Type="LED" Lumens="0" />
-   <Dimensions Width="335" Height="182" Depth="93" Weight="3.25" />
-   <Lens Name="Other" DegreesMin="75" DegreesMax="75" />
-   <Focus Type="Fixed" TiltMax="0" PanMax="0" />
-   <Technical DmxConnector="3-pin" PowerConsumption="130" />
+   <Bulb ColourTemperature="0" Type="LED" Lumens="0"/>
+   <Dimensions Width="335" Height="182" Depth="93" Weight="3.25"/>
+   <Lens Name="Other" DegreesMin="75" DegreesMax="75"/>
+   <Focus Type="Fixed" TiltMax="0" PanMax="0"/>
+   <Technical DmxConnector="3-pin" PowerConsumption="130"/>
   </Physical>
   <Channel Number="0">Dimmer</Channel>
   <Channel Number="1">Strobe 3</Channel>
   <Channel Number="2">Programs</Channel>
@@ -124,13 +124,13 @@
   <Channel Number="4">Sound Control</Channel>
  </Mode>
  <Mode Name="7-channel">
   <Physical>
-   <Bulb ColourTemperature="0" Type="LED" Lumens="0" />
-   <Dimensions Width="335" Height="182" Depth="93" Weight="3.25" />
-   <Lens Name="Other" DegreesMin="75" DegreesMax="75" />
-   <Focus Type="Fixed" TiltMax="0" PanMax="0" />
-   <Technical DmxConnector="3-pin" PowerConsumption="130" />
+   <Bulb ColourTemperature="0" Type="LED" Lumens="0"/>
+   <Dimensions Width="335" Height="182" Depth="93" Weight="3.25"/>
+   <Lens Name="Other" DegreesMin="75" DegreesMax="75"/>
+   <Focus Type="Fixed" TiltMax="0" PanMax="0"/>
+   <Technical DmxConnector="3-pin" PowerConsumption="130"/>
   </Physical>
   <Channel Number="0">Dimmer</Channel>
   <Channel Number="1">Strobe 2</Channel>
   <Channel Number="2">Flash Duration</Channel>

```
```diff
Index: lightmaxx/vega-zoom-wash.qxf
===================================================================
--- lightmaxx/vega-zoom-wash.qxf
+++ lightmaxx/vega-zoom-wash.qxf
@@ -108,13 +108,13 @@
   <Capability Min="226" Max="255">no function</Capability>
  </Channel>
  <Mode Name="16-channel">
   <Physical>
-   <Bulb ColourTemperature="0" Type="LED" Lumens="0" />
-   <Dimensions Width="260" Height="280" Depth="340" Weight="4.5" />
-   <Lens Name="Other" DegreesMin="6" DegreesMax="45" />
-   <Focus Type="Head" TiltMax="190" PanMax="540" />
-   <Technical DmxConnector="3-pin" PowerConsumption="95" />
+   <Bulb ColourTemperature="0" Type="LED" Lumens="0"/>
+   <Dimensions Width="260" Height="280" Depth="340" Weight="4.5"/>
+   <Lens Name="Other" DegreesMin="6" DegreesMax="45"/>
+   <Focus Type="Head" TiltMax="190" PanMax="540"/>
+   <Technical DmxConnector="3-pin" PowerConsumption="95"/>
   </Physical>
   <Channel Number="0">Pan</Channel>
   <Channel Number="1">Pan fine</Channel>
   <Channel Number="2">Tilt</Channel>

```

FloEdelmann commented:
*[Updated plugin output diff comment](https://github.com/FloEdelmann/open-fixture-library/pull/110#issuecomment-299043260)*

FloEdelmann commented:
*[Updated plugin output diff comment](https://github.com/FloEdelmann/open-fixture-library/pull/110#issuecomment-299043260)*

fxedel commented:
@FloEdelmann If you look at the latest [plugin output diff](#issuecomment-299043260), you'll see that using the xml builder will not make this xml code:
```xml
<parent>
  <a/>
  <b/>
  <b>Bla</b>
  <a>Blubb</a>
  <b/>
</parent>
```
but this:
```xml
<parent>
  <a/>
  <a>Blubb</a>
  <b/>
  <b>Bla</b>
  <b/>
</parent>
```

That's because you define children using arrays for each element type and there's no way defining an order between different element types.
```js
let xml = {
  parent: {
    a: [
      {},
      'Blubb'
    ],
    b: [
      {},
      'Bla',
      {}
    ]
  }
}
```

That's why e:cue's channels are outputted not ordered by channel index but by element type (`ChannelBeam`, `ChannelIntensity`, ...). **So, is the channel order important in e:cue?** There's still the `ClassicPos` attribute that clearly defines the dmx channel position.

FloEdelmann commented:
I guess it is not important since there are the `DmxByte0` and `DmxByte1` attributes (the `ClassicPos` attribute is irrelevant here) but I'd still rather have it in the right order to improve readability of the output file.

FloEdelmann commented:
*[Updated plugin output diff comment](https://github.com/FloEdelmann/open-fixture-library/pull/110#issuecomment-299043260)*

FloEdelmann commented:
*[Updated plugin output diff comment](https://github.com/FloEdelmann/open-fixture-library/pull/110#issuecomment-299043260)*

FloEdelmann commented:
*[Updated plugin output diff comment](https://github.com/FloEdelmann/open-fixture-library/pull/110#issuecomment-299043260)*

fxedel commented:
@FloEdelmann Please already have a look at this. I'll remove the dummy fixtures and add a real fixture using the switched channels, but except of that, I see that PR as finished now.

FloEdelmann commented:
*[Updated plugin output diff comment](https://github.com/FloEdelmann/open-fixture-library/pull/110#issuecomment-299043260)*

FloEdelmann commented:
Please fix the outstanding Codacy issues and use a real fixture as you said, otherwise I think it's good to go. I'll leave a final review then.

FloEdelmann commented:
*[Updated plugin output diff comment](https://github.com/FloEdelmann/open-fixture-library/pull/110#issuecomment-299043260)*

fxedel commented:
* Fixed (capabilities would be nice, but we don't know which DMX value represents which hue)
* Also fixed. We should think about an auto-generated name "[Switched channel 1] / [Switched channel 2] / ..." or something like that
* The manual is bilingual

FloEdelmann commented:
*[Updated plugin output diff comment](https://github.com/FloEdelmann/open-fixture-library/pull/110#issuecomment-299043260)*

fxedel commented:
It might also be reasonable naming the exported default channel "Program Speed / Mic Sensitivity" – however, it interferes with the channel's data (type, capabilities, etc.) with refers to "Program Speed".

FloEdelmann commented on fixtures/schema.js:
```diff
@@ -49,8 +51,10 @@ var Physical = schema({
   '*': Function
 });
 
+var Range = Array.of(2, DMXValue);
```
Please rename this `DMXRange`, since we also have a range in `degreesMinMax`.

FloEdelmann commented on fixtures/schema.js:
```diff
@@ -75,6 +79,15 @@ var Channel = schema({
+  '?name': [String, null], // null: use channel key
+  'trigger': ChannelKey,
+  'channels': schema({
+    '*': [Range, Array.of(1, Infinity, Range)] // '*' is the channel key of an switched channel
```
An array of DMXRanges should at least contain two of them. Or should we just always allow arrays of ranges?

FloEdelmann commented on fixtures/schema.js:
```diff
@@ -75,6 +79,15 @@ var Channel = schema({
 
+var SwitchingChannel = schema({
+  '?name': [String, null], // null: use channel key
+  'trigger': ChannelKey,
```
Rename this to `triggerChannel`.

FloEdelmann commented on tests/fixture_valid.js:
```diff
@@ -84,12 +85,25 @@ module.exports.checkFixture = function checkFixture(fixture, usedShortNames=[])
-        if (!(ch in fixture.availableChannels) && ch !== null) {
+        const isNormalChannel = ch in fixture.availableChannels;
+        const isSwitchingChannel = 'switchingChannels' in fixture && ch in fixture.switchingChannels;
+        if (!isNormalChannel &&
```
this fits in one line

FloEdelmann commented on tests/fixture_valid.js:
```diff
@@ -125,7 +139,14 @@ module.exports.checkFixture = function checkFixture(fixture, usedShortNames=[])
 
     for (const ch in fixture.availableChannels) {
-      if (usedChannels.indexOf(ch) == -1) {
+      isUsed: if (!usedChannels.includes(ch)) {
```
Can we refactor this to not having to use a label?

FloEdelmann commented on tests/fixture_valid.js:
```diff
@@ -194,6 +215,99 @@ module.exports.checkFixture = function checkFixture(fixture, usedShortNames=[])
+                    else {
+                      usedRanges.splice(i, 0, range);
+                    }
+                    return;
```
return or break?

fxedel replied:
Actually the whole range overlapping test was wrong.

FloEdelmann commented on tests/fixture_valid.js:
```diff
@@ -194,6 +215,99 @@ module.exports.checkFixture = function checkFixture(fixture, usedShortNames=[])
+          }
+
+          for (let i = -1; i < usedRanges.length; i++) {
+            let previousRange = i === -1 ?
```
those also fit in one line

FloEdelmann commented:
*[Updated plugin output diff comment](https://github.com/FloEdelmann/open-fixture-library/pull/110#issuecomment-299043260)*

FloEdelmann commented:
*[Updated plugin output diff comment](https://github.com/FloEdelmann/open-fixture-library/pull/110#issuecomment-299043260)*

FloEdelmann commented:
Please also update *american-dj/revo-4-ir* to support its switching channel correctly.

fxedel commented:
I still need to push the new syntax with Aliases (like in #122) as soon as #122 is merged (cause there are a lot of merge conflicts to resolve). But thanks for the notice.

FloEdelmann commented:
*[Updated plugin output diff comment](https://github.com/FloEdelmann/open-fixture-library/pull/110#issuecomment-299043260)*

FloEdelmann commented on tests/fixture_valid.js:
```diff
@@ -132,8 +146,20 @@ module.exports.checkFixture = function checkFixture(fixture, usedShortNames=[])
-      if (usedChannels.indexOf(ch) === -1) {
-        result.warnings.push(`Channel '${ch}' defined but never used.`);
+      if (!usedChannels.includes(ch)) {
+        let isUsed = false;
```
Please condense this into one statement: `const isUsed = 'switchingChannels' in fixture && Object.keys(fixture.switchingChannels).some(key => ...);`

FloEdelmann commented on views/pages/single_fixture.js:
```diff
@@ -141,12 +141,27 @@ module.exports = function(options) {
-      const channel = fixture.availableChannels[ch];
+      const isSwitchingChannel = 'switchingChannels' in fixture && ch in fixture.switchingChannels;
+
+      let channel;
```
why not use `x ? y : z` here?

FloEdelmann commented on views/pages/single_fixture.js:
```diff
@@ -376,11 +389,11 @@ function handleChannel(channel) {
       str += '</td>';
 
-      str += `<td class="capability-color" title="color: ${cap.color}">`;
+      str += '<td class="capability-color" title="color: ${cap.color}">';
```
wrong. This was possibly a merge mistake.

FloEdelmann commented on views/pages/single_fixture.js:
```diff
@@ -376,11 +389,11 @@ function handleChannel(channel) {
       str += '</td>';
 
-      str += `<td class="capability-color2" title="color 2: ${cap.color2}">`;
+      str += '<td class="capability-color2" title="color 2: ${cap.color2}">';
```
Again.

FloEdelmann commented on plugins/qlcplus.js:
```diff
@@ -137,7 +170,7 @@ function exportHandleAvailableChannels(fixture, defaults) {
     };
 
-    if (chData.type === 'Intensity') {
+    if (chData.type == 'Intensity') {
```
Merge mistake

FloEdelmann commented on plugins/ecue.js:
```diff
@@ -116,123 +116,157 @@ function exportHandleModes(fixture, defaults, physical, xmlMan) {
-  for (let channelList of fixture.multiByteChannels) {
-    if (channelList.indexOf(channelKey) !== -1) {
-      return channelList;
+  function getCorrespondingMultiByteChannels(channelKey, fixture) {
```
`fixture` is not needed as a parameter here

FloEdelmann commented on plugins/ecue.js:
```diff
@@ -116,123 +116,157 @@ function exportHandleModes(fixture, defaults, physical, xmlMan) {
+      const ranges = switchingChannel.channels[switchedChannelKey];
+
+      if (!Array.isArray(ranges[0])) {
+        if (ranges[0] <= defaultValue && defaultValue <= ranges[1]) {
```
you could do `ranges = [ranges];` here and just fall through into the other case to reduce code duplication.

FloEdelmann commented:
*[Updated plugin output diff comment](https://github.com/FloEdelmann/open-fixture-library/pull/110#issuecomment-299043260)*

FloEdelmann commented:
*[Updated plugin output diff comment](https://github.com/FloEdelmann/open-fixture-library/pull/110#issuecomment-299043260)*

FloEdelmann commented:
*[Updated plugin output diff comment](https://github.com/FloEdelmann/open-fixture-library/pull/110#issuecomment-299043260)*

fxedel commented:
I just reset the branch to master and implemented the new syntax directly. Most things are finished except of UI.

FloEdelmann commented:
*[Updated plugin output diff comment](https://github.com/FloEdelmann/open-fixture-library/pull/110#issuecomment-299043260)*

FloEdelmann commented:
*[Updated plugin output diff comment](https://github.com/FloEdelmann/open-fixture-library/pull/110#issuecomment-299043260)*

FloEdelmann commented:
*[Updated plugin output diff comment](https://github.com/FloEdelmann/open-fixture-library/pull/110#issuecomment-299043260)*

FloEdelmann commented on fixtures/schema.js:
```diff
@@ -51,31 +51,33 @@ var Physical = schema({
   'type': ['Intensity', 'Strobe', 'Shutter', 'Speed', 'SingleColor', 'MultiColor', 'Gobo', 'Prism', 'Pan', 'Tilt', 'Beam', 'Effect', 'Maintenance', 'Nothing'],
   '?color': ['Red', 'Green', 'Blue', 'Cyan', 'Magenta', 'Yellow', 'Amber', 'White', 'UV', 'Lime', 'Indigo'],
   '?fineChannelAliases': Array.of(1, Infinity, ChannelAliasKey),
+  '?switchesChannels': Array.of(1, Infinity, ChannelAliasKey), // e.g. =["Strobe/Pan"] and switchToChannels="Strobe" or "Pan" in this channel's capabilities: Use "Strobe/Pan" in a mode, it's behavior switches between "Strobe" and "Pan", depending on this channel's value
```
The comment is a bit lengthy, also it's `its` not `it's` here.

FloEdelmann commented on fixtures/schema.js:
```diff
@@ -51,31 +51,33 @@ var Physical = schema({
   '?color': Color,
   '?color2': Color,
   '?image': NonEmptyString,
+  '?switchToChannels': Array.of(1, Infinity, ChannelKey), // same array length as Channel.switchesChannels
```
I still think an object would be better here:

```
"switchChannels": {
  "Chase Speed / Mic Sensitivity": "Chase Speed"
}
```

Then it would be clear which switching channel would be set to which actual channel. An array does not yield the same intuitiveness.

fxedel replied:
So you wouldn't declare switching channels in the channel itself but only in each capability? But then we had to write the same switching channel alias n times in n capabilities which is redundant and therefore leads to mistakes and worse maintainability.

Though, I agree that it would be more intuitive and more clear.

FloEdelmann commented on plugins/qlcplus.js:
```diff
@@ -19,11 +19,14 @@ module.exports.export = function exportQLCplus(library, options) {
     for (const mode of fixture.modes) {
       for (let i=0; i<mode.channels.length; i++) {
         if (mode.channels[i] == null) {
-          mode.channels[i] = 'No Function ' + i;
```
You can not reuse the same channel more than once in a mode! Does the new code account for that?

fxedel replied:
Why shouldn't it be possible to reuse the same channel more than once in a mode? But I admit I didn't thought about that case and also didn't test it.

FloEdelmann replied:
It is simply not possible in QLC+, don't ask me why.

FloEdelmann commented on tests/fixture_valid.js:
```diff
@@ -147,25 +66,43 @@ module.exports.checkFixture = function checkFixture(fixture, usedShortNames=[])
+
+        if (!('defaultValue' in channel)) {
+          result.errors.push({
+            description: `defaultValue is missing in channel '${ch}' as it defines switching channels.`,
```
`although` instead of `as`

FloEdelmann commented on tests/fixture_valid.js:
```diff
@@ -192,47 +129,41 @@ module.exports.checkFixture = function checkFixture(fixture, usedShortNames=[])
               description: `range values must be strictly less than ${dmxMaxBound} in capability #${i} in channel '${ch}'.`,
               error: null
             });
-            break;
```
If one range is invalid, the later tests can not rely on it anymore and are therefore not meaningful (and in the worst case misleading). So better just exit the test and potentially hide some errors that will then only occur on retesting.

fxedel replied:
If the first range is invalid, there's no use to make the other range tests. That's why we simply use `else if`. But it doesn't make sense to skip color / image / switching channel tests only because the range is wrong.

FloEdelmann replied:
It also does not make sense to test the next capability's range, that's why we break out of the capability loop.

FloEdelmann commented on tests/fixture_valid.js:
```diff
@@ -192,47 +129,41 @@ module.exports.checkFixture = function checkFixture(fixture, usedShortNames=[])
               description: `range invalid in capability #${i} in channel '${ch}'.`,
               error: null
             });
-            break;
```
Again.

FloEdelmann commented on tests/fixture_valid.js:
```diff
@@ -192,47 +129,41 @@ module.exports.checkFixture = function checkFixture(fixture, usedShortNames=[])
               description: `ranges overlapping in capabilities #${i-1} and #${i} in channel '${ch}'.`,
               error: null
             });
-            break;
```
Again

FloEdelmann commented on tests/fixture_valid.js:
```diff
@@ -192,47 +129,41 @@ module.exports.checkFixture = function checkFixture(fixture, usedShortNames=[])
                 description: `ranges overlapping when used in coarse channel only mode in capabilities #${i-1} and #${i} in channel '${ch}'.`,
                 error: null
               });
-              break;
```
Again

FloEdelmann commented on tests/fixture_valid.js:
```diff
@@ -246,14 +177,165 @@ module.exports.checkFixture = function checkFixture(fixture, usedShortNames=[])
+
+            // switched channels are used with the switching channel and don't need to be used somewhere else
+            for (const switchedChannel of cap.switchToChannels) {
+              if (!(switchedChannel in fixture.availableChannels)) {
```
Could it also be a fine channel?

fxedel replied:
I didn't allow switch fine channels by purpose as I don't think that it makes much sense and it only makes things even more complicated. What do you think about it?

However, we can still add the functionality as soon as we see a fixture that switches fine channels.

FloEdelmann replied:
Yep, that will suffice.

FloEdelmann commented on tests/fixture_valid.js:
```diff
@@ -246,14 +177,165 @@ module.exports.checkFixture = function checkFixture(fixture, usedShortNames=[])
+        if (ch in fineChannels) {
+          if (!mode.channels.includes(fineChannels[ch])) {
+            result.errors.push({
+              description: `mode '${modeShortName}' uses switching channel '${ch}' (#${chIndex}) but is missing its trigger channel '${fineChannels[ch]}'`,
```
wrong error description

fxedel replied:
Oops.

fxedel replied:
Currently, the test only makes sure that the coarse counterpart of a fine channel is used as well. But if a mode uses an 8bit and a 24bit channel, we should ensure that it also contains the 16bit channel, shouldn't we?

FloEdelmann replied:
Possibly

FloEdelmann commented on tests/fixture_valid.js:
```diff
@@ -246,14 +177,165 @@ module.exports.checkFixture = function checkFixture(fixture, usedShortNames=[])
+        const head = fixture.heads[key];
+
+        for (let i = 0; i < head.length; i++) {
+          if (!fixture.availableChannels[head[i]]) {
```
Can it also be a fine channel?

fxedel replied:
Definitely yes. (Even though I don't think that I actively removed this functionality)

FloEdelmann commented on views/pages/single_fixture.js:
```diff
@@ -413,6 +467,15 @@ function handleChannel(channel, mode) {
+          const switchingChannel = channel.switchesChannels[i];
+          const switchingChannelIndex = mode.channels.indexOf(switchingChannel);
+          const switchTo = cap.switchToChannels[i];
+          str += `<tr><td colspan="7" class="switch-to-channel"><small>${switchingChannel} (channel ${switchingChannelIndex}) → ${switchTo}</small></td></tr>`;
```
why a `<small>` tag when you already have a CSS class where you can change the font-size?

FloEdelmann commented:
<!-- github-plugin-diff -->
<!-- commit = a37a153c34df2d14c06eb91a3276d9dfe61af981 -->
Last updated: 6/11/2017, 1:54:17 PM
## Diff plugin outputs test
### Modified `fixtures/american-dj/revo-4-ir.json` in this PR
#### Output changed for plugin `ecue`
```diff
Index: UserLibrary.xml
===================================================================
--- UserLibrary.xml
+++ UserLibrary.xml
@@ -23,9 +23,9 @@
                         <Range Name="Pattern / Chase 10" Start="194" End="214" AutoMenu="1" Centre="0"/>
                         <Range Name="Pattern / Chase 11" Start="215" End="235" AutoMenu="1" Centre="0"/>
                         <Range Name="Pattern / Chase 12" Start="236" End="255" AutoMenu="1" Centre="0"/>
                     </ChannelBeam>
-                    <ChannelColor Name="Color" DefaultValue="0" Highlight="0" Deflection="0" DmxByte0="3" DmxByte1="0" Constant="0" Crossfade="0" Invert="0" Precedence="LTP" ClassicPos="3">
+                    <ChannelColor Name="Color / Chase Speed" DefaultValue="0" Highlight="0" Deflection="0" DmxByte0="3" DmxByte1="0" Constant="0" Crossfade="0" Invert="0" Precedence="LTP" ClassicPos="3">
                         <Range Name="R" Start="0" End="17" AutoMenu="1" Centre="0"/>
                         <Range Name="G" Start="18" End="34" AutoMenu="1" Centre="0"/>
                         <Range Name="B" Start="35" End="51" AutoMenu="1" Centre="0"/>
                         <Range Name="W" Start="52" End="68" AutoMenu="1" Centre="0"/>

```
#### Output changed for plugin `qlcplus`
```diff
Index: american-dj/revo-4-ir.qxf
===================================================================
--- american-dj/revo-4-ir.qxf
+++ american-dj/revo-4-ir.qxf
@@ -1593,8 +1593,26 @@
   <Colour>Red</Colour>
   <Capability Min="0" Max="127">off</Capability>
   <Capability Min="128" Max="255">on</Capability>
  </Channel>
+ <Channel Name="Color / Chase Speed">
+  <Group Byte="0">Colour</Group>
+  <Capability Min="0" Max="17" Color="#ff0000">R</Capability>
+  <Capability Min="18" Max="34" Color="#00ff00">G</Capability>
+  <Capability Min="35" Max="51" Color="#0000ff">B</Capability>
+  <Capability Min="52" Max="68" Color="#ffffff">W</Capability>
+  <Capability Min="69" Max="85" Color="#ffff00">R+G</Capability>
+  <Capability Min="86" Max="102" Color="#ff00ff">R+B</Capability>
+  <Capability Min="103" Max="119" Color="#ffaaaa">R+W</Capability>
+  <Capability Min="120" Max="136" Color="#00ffff">G+B</Capability>
+  <Capability Min="137" Max="153" Color="#aaffaa">G+W</Capability>
+  <Capability Min="154" Max="170" Color="#aaaaff">B+W</Capability>
+  <Capability Min="171" Max="187">R+G+B</Capability>
+  <Capability Min="188" Max="204">R+G+W</Capability>
+  <Capability Min="205" Max="221">R+B+W</Capability>
+  <Capability Min="222" Max="238">G+B+W</Capability>
+  <Capability Min="239" Max="255">R+G+B+W</Capability>
+ </Channel>
  <Mode Name="4 channel">
   <Physical>
    <Bulb ColourTemperature="0" Type="Other" Lumens="0"/>
    <Dimensions Width="312" Height="334" Depth="304" Weight="4.9"/>
@@ -1603,9 +1621,9 @@
    <Technical DmxConnector="3-pin" PowerConsumption="34"/>
   </Physical>
   <Channel Number="0">Function</Channel>
   <Channel Number="1">Pattern / Chase</Channel>
-  <Channel Number="2">Color</Channel>
+  <Channel Number="2">Color / Chase Speed</Channel>
   <Channel Number="3">Strobe</Channel>
  </Mode>
  <Mode Name="256 channel">
   <Physical>

```
### Modified plugin `ecue` in this PR
Plugins are always tested with the following fixtures:
 - fixtures/cameo/thunder-wash-600-w.json
 - fixtures/eurolite/led-tmh-18.json
Output files not changed.
### Modified plugin `qlcplus` in this PR
Plugins are always tested with the following fixtures:
 - fixtures/cameo/thunder-wash-600-w.json
 - fixtures/eurolite/led-tmh-18.json
Output files not changed.

FloEdelmann commented on plugins/qlcplus.js:
```diff
@@ -17,12 +17,13 @@ module.exports.export = function exportQLCplus(library, options) {
-          mode.channels[i] = 'No Function';
+      for (let i = 0; i < mode.channels.length; i++) {
+        if (mode.channels[i] === null) {
+          const noFunctionChKey = `No Function ${mode.name} ${i}`;
```
Why mode.name? No Function 0 can be used in more than one channel.

fxedel replied:
`mode.name` reduces the probability that there is already a channel with that key – if the fixture has a `No function` and a `No function 1` channel, then a `null` at index 1 would use the existing `No function 1` channel which can have unintended side effects.

FloEdelmann commented:
*[Updated plugin output diff comment](https://github.com/FloEdelmann/open-fixture-library/pull/110#issuecomment-307630451)*

FloEdelmann commented:
*[Updated plugin output diff comment](https://github.com/FloEdelmann/open-fixture-library/pull/110#issuecomment-307630451)*
